Readers/src/RecoJetReaderAK4.cc

```diff
@@ -341,17 +336,12 @@ RecoJetReaderAK4::read() const
         qgl,
         gInstance->jet_bRegCorr_[idxJet],
         gInstance->jet_bRegRes_[idxJet],
-        gInstance->jet_pullEta_[idxJet],
-        gInstance->jet_pullPhi_[idxJet],
-        gInstance->jet_pullMag_[idxJet],
         jet_id,
         gInstance->jet_puId_[idxJet],
         gInstance->jet_genMatchIdx_[idxJet],
-        gInstance->jet_jetIdx_[idxJet],
```

Comment:
I discussed with Christian about this and he told that this jetIdx is used while cleaning the jets and currently we do cleaning by dr. If we need this we can add them in future but currently reduce the dependency of post-processing step

Reply:
Hi, 

I realized I may have made a mistake regarding this point. Karl, can you remind me in which analyzes we use jet cleaning by dR and in which we use cleaning by index ? 
I have a related question: do we compute the index that we use for the cleanining ourselves in the postprocessing or do we use the default nanoAOD implementation in
  https://github.com/HEP-KBFI/cmssw/blob/master/PhysicsTools/NanoAOD/python/nano_cff.py#L32-L38
I agree with Saswati that it is good to trim down as much as possible the postprocessing code, but if we can use the index computed by default in nanoAOD for the cleaning then I would vote for keeping the jet cleaning by index.

Reply:
We use the index-based cleaning in ttH and HH-multilepton analysis, and dR-based cleaning in HH-bbWW. The indices to the jet collection are determined during NanoAOD production. The only task from our part is to actually read the indices that are attributed to the selected reco leptons and taus, find the corresponding jets and remove the jets.

Index-based cleaning is less indiscriminate, since it cleans up to one jet per lepton, while dR-based cleaning can remove multiple jets per lepton. If it's not possible to keep both methods and switch between either if needed, then I propose we stick to the index-based approach. That's what other groups also seem to do, at least in ttH analysis. And unlike with HH-multilepton, we definitely need to collaborate and synchronize with other groups on the event selection.

Reply:
I see some strange behavior. Though jetCleaningbyIndex is set True here
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Framework/test/produceNtuple_cfg.py#L263 But in the code it is always considering this
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/ParticleCollectionCleaner.h#L27 instead of this      https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h#L18

Reply:
According to the code that's currently on github, I don't see how this is possible. Please check again (grep for eg jetCleaningByIndex), recompile the code, and/or try adding debugging statements to make sure that the `EventReader` instance is correctly set up.

Reply:
there is some problem in the code as in my current tree there is no branch Jet_jetIdx but still the code works fine. Even I removed all jetIdx variables from the code but still it works. But I'll add more debugging in the code and see later

Reply:
The `jetIdx` variables are available by default in NanoAOD, except for the jets themselves. This is because when we read the jet collection, the jets are already ordered as they're stored in NanoAOD. This is was not guaranteed in the old FW, where we had to rely on [`jetIdx` module](https://github.com/HEP-KBFI/tth-nanoAOD-tools/blob/master/python/postprocessing/modules/jetIdxProducer.py) that stored the original position of each jet in `Jet_jetIdx` branch.

Reply:
I bring the jetIdx back here https://github.com/HEP-KBFI/TallinnNtupleProducer/commit/a4208400d6df3804106fb81f27bf0af61469d057 but I am still not able to find out why it is taking PartcleCollectionCleaner operator for AK4 jets. I took print out in main branch but it prints ParticleCollectionCleaner.May be I am doing some silly mistake somewhere

Reply:
Hi Saswati,

there is a special piece of code that we use for cleaning AK4 jets by index:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h

Tehnically, this class inherits from the class for general dR cleaning, which is defined here:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/ParticleCollectionCleaner.h

We switch between cleaning AK4 jets by dR and cleaning by index using these lines of code:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Readers/src/EventReader.cc#L186-L193

I hope that helps,

Christian

Reply:
Hi @veelken ,

 I understand this, I already looked into this before but didn't understand why it enters the operator [here](https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/ParticleCollectionCleaner.h#L27) when it is called for jetIndexMatching instead of [here](https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h#L18). So for cross check I requested Laurits to check and he also told me that it never enters the operator inside RecoJetCollectionCleanerByIndex. I request someone who are using main branch to check this. May be I made some changes in my branch which I can't indentify.

Reply:
So I had a look at the issue myself and I confirm that it does not work. The reason being, we define `jetCleanerAK4_dR04_` in terms of base class:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/fa246726febeed601a0ca88bc4471bfd862b65cb/Readers/interface/EventReader.h#L144
And then we call the `operator()` function, which naturally points to the base class:
https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/fa246726febeed601a0ca88bc4471bfd862b65cb/Readers/src/EventReader.cc#L563-L565
The natural solution here would be to promote the base class function to `virtual`, so that the ambiguity could be resolved at runtime, but C++ forbids declaring template functions as `virtual`. For a minimal working example, see https://ideone.com/OLqPyn and try changing the template function in the base class to `virtual`.

I don't have a good solution other than to go back to what I initially implemented in the old FW: that we have a dedicated class for `RecoJetCollectionCleanerByIndexAK4` that does not inherit from `ParticleCollectionCleaner`, create two separate instances of the cleaning classes, and resolve ambiguity between them with a simple if-else clause / ternary expression in `EventReader`.

By the way, I would also move the cleaning classes to stack instead of having them on heap. I don't see a clear benefit in declaring pointers and then having to call `->operator()(...)` explicitly.

Reply:
Hi,

I agree with Karl's suggestion. Saswati, please change 
  https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/main/Cleaners/interface/RecoJetCollectionCleanerByIndexAK4.h#L8
to
  https://github.com/HEP-KBFI/tth-htt/blob/master/interface/RecoJetCollectionCleanerByIndex.h#L4
and then use lines like
  https://github.com/HEP-KBFI/tth-htt/blob/master/bin/analyze_2lss_1tau.cc#L1337-L1340
to switch between jet cleaning by dR and jet cleaning by index.

I also agree with Karl's comment that we can replace
  RecoJetCollectionCleanerAK4 * jetCleanerAK4_dR04_; // heap
by
  RecoJetCollectionCleanerAK4 jetCleanerAK4_dR04_; // stack
in https://github.com/HEP-KBFI/TallinnNtupleProducer/blob/fa246726febeed601a0ca88bc4471bfd862b65cb/Readers/interface/EventReader.h#L144-L145.
The reason why I used pointers is that I can declare a base-class pointer and assign this pointer either to the base-class (RecoJetCollectionCleanerAK4) or to the derived class (RecoJetCollectionCleanerByIndexAK4).

Reply:
done [here](https://github.com/HEP-KBFI/TallinnNtupleProducer/pull/3/commits/d0bddf50ec96d8fb3edf7767a44bc6efd8d2d829) . I think we can replace selector pointer also by object, right?

Reply:
In principle, yes, but I would keep the changes as minimal as possible at this point because this PR is huge already.